pages/directory/index.tsx

```diff
@@ -1,28 +1,106 @@
-import { Box, Container, List, ListItem, Typography } from '@mui/material';
+import { Box, Container, CircularProgress, List, ListItem, Typography } from '@mui/material';
+import axios from 'axios';
 import { NextPage } from 'next';
-import React from 'react';
+import * as React from 'react';
 
+import ErrorPopup from 'client/components/ErrorPopup';
 import createAxiosInstance from 'client/lib/axios';
 import { ClientUser } from 'shared/User';
 
-const Directory: NextPage<{ users: ClientUser[]; error: string }> = props => {
-  const { users, error } = props;
+const Directory: NextPage<{ users: ClientUser[]; totalPages: number; error: string }> = props => {
+  const { users, totalPages, error } = props;
+  const lastCardRef = React.useRef(null);
+  const [userData, setUserData] = React.useState(users);
+  const [currentPage, setCurrentPage] = React.useState(1);
+  const [totalPageNum, setTotalPageNum] = React.useState(totalPages);
+  const [isLoading, setIsLoading] = React.useState(false);
+  const [showError, setShowError] = React.useState(false);
+  const [errorMessage, setErrorMessage] = React.useState('');
+
+  async function fetchUsers(page: number) {
+    try {
+      const response = await axios.get(`/api/users?page=${page}`);
+      return {
+        page: response.data.page,
+        total: response.data.totalPages,
+        users: response.data.users,
+      };
+    } catch (err) {
+      if (err.code === 'ERR_BAD_REQUEST') {
+        throw new Error('Page out of range', { cause: err.message });
+      }
+    }
+  }
+
+  React.useEffect(() => {
+    const height = getElementHeight(document.querySelector('#last-card'));
+
+    function getElementHeight(el: HTMLElement) {
+      const rect = el.getBoundingClientRect();
+      const { marginTop, marginBottom } = getComputedStyle(el);
+
+      return rect.height + parseFloat(marginTop) + parseFloat(marginBottom);
+    }
+
+    const intersectionOptions = {
+      // root default = document's viewport
+      rootMargin: `${height}px`,
+      threshold: 0,
+    };
+
+    const lastCardObserver = new IntersectionObserver(async ([lastCard]) => {
+      // guard clause, On mount intersection triggers w/ false, should intersect only once
+      if (!lastCard.isIntersecting) return;
+
+      lastCardObserver.disconnect();
+      // stop loading users when at max pages before fetching next page
+      if (currentPage + 1 > totalPageNum) {
+        return;
+      }
+      setIsLoading(true);
+      try {
+        const { page, total, users } = await fetchUsers(currentPage + 1);
+        setCurrentPage(page);
+        setTotalPageNum(total);
+        setUserData([...userData, ...users]);
+      } catch (err) {
+        if (err.message === 'Page out of range') {
```

Comment:
I'm not sure this conditional is necessary... or, if kept - should we provide an `else` condition with a generic error?